### PR TITLE
add interface to define and query channel groups

### PR DESCRIPTION
**Pull request by cg-laser**
the first step in making NuRadioReco ready for LOFAR

### CODE REVIEW

**Review by cg-laser — COMMENTED** [review]
one thing that I noticed which makes the PR difficult to review are the many many style changes, mostly concerning maximum line length. It would be good to avoid that. Our coding convention is to NOT stay within 80 characters. Nowadays, everyone has large screens and this restriction is not required anymore. And expanding a simple warning printout over three lines makes the code less readable. 
I guess this comes from some autoformatter setting. Please deactivate that or have stylistic changes to a different PR.

**Comment by sjoerd-bouma** [comment]
> one thing that I noticed which makes the PR difficult to review are the many many style changes, mostly concerning maximum line length. It would be good to avoid that. Our coding convention is to NOT stay within 80 characters. Nowadays, everyone has large screens and this restriction is not required anymore. And expanding a simple warning printout over three lines makes the code less readable. I guess this comes from some autoformatter setting. Please deactivate that or have stylistic changes to a different PR.

@cg-laser I think this PR currently shows some changes that have already been merged into develop, not sure why. Maybe this will be resolved by merging in the latest develop, otherwise we should close and reopen the PR.

**Comment by MijnheerD** [comment]
@cg-laser My apologies, I think one (more or less) automatic refactoring of the code made it through one of my commits. I understand this makes it difficult to review. Only `eventWriter.py` should be affected by this though, all other files also contain substantial changes/additions.

However, I do not understand why we should not adhere to PEP8. It is a Python standard after all, and is still useful in the age of big screens in my opinion. Last week for example, while I was working on this branch, I found myself often working with two files open in split-screen. In this case having a maximum line length of 80 characters is very helpful. Though I have no strong feelings about going till 120 characters, but longer quickly becomes unreadable in my experience.

File: NuRadioReco/modules/io/eventWriter.py
Comment (cg-laser) [inline NuRadioReco/modules/io/eventWriter.py:296]: can you add a comment explaining what the decorator does and why it is added?

**Review by cg-laser — COMMENTED** [review]
the PR looks good to me. I only made a few smaller in-line comments. 

I didn't check any of the LOFAR modules though.

File: NuRadioReco/modules/io/LOFAR/readLOFARData.py
Comment (sjoerd-bouma) [inline NuRadioReco/modules/io/LOFAR/readLOFARData.py:454]: I think this came up in the call - this looks a bit risky to me, probably safer/more natural to implement this in the detector descriptions. E.g. what if a station is slightly differently oriented, or you end up calibrating the antenna orientations so they're no longer exactly 225 degrees, etc.

**Review by sjoerd-bouma — COMMENTED** [review]
A couple of things
- I tend to leave very nitpicky reviews - please treat most/all of my comments as suggestions, not as 'this should change before you can merge'!
- Like Christian I did not go over the LOFAR modules particularly carefully, I just flagged a couple of things that caught my eye.

In terms of students starting work before this is merged into develop - feel free to create another branch that's based off this one, which should allow parallel development to some extent.

**Comment by sjoerd-bouma** [comment]
Also re channel group ids - I may be mistaken, but I think Christian seems to expect them to refer to groups of (for LOFAR) dual-polarized antennas (e.g. 1,...48 for the 48 x 2 antennas in a LOFAR station), whereas you have used them to refer to the groups of a single polarization (i.e. 0 for 225 deg orientation and 1 for the 135 deg orientation). 

From my side, I originally expected the feature to do the former, but I do not have a strong preference for which of the two concepts the channel group id should represent (though for example assigning Efields to each group of dual-polarized antennas from a CoREAS file would be more straightforwardly implemented in the first case). I do think it's worth thinking through and documenting clearly somewhere.

**Comment by MijnheerD** [comment]
Thanks a lot for your comments Sjoerd and Christian! I will have a look at them as soon possible and fix the issues you mentioned.

Concerning the channel group ids, I indeed interpreted those as polarisations rather than "antenna index" cause in the calibration phase we need the polarisation of the channel in order to choose which curve to pick. However, it seems that at least for high levels analyses/applications (like the CoREAS reader for example) it would be more beneficial to use the id to refer to antennas instead. For LOFAR we will anyway do both things (i.e. grab all channels with the same polarisations and grab the two channels corresponding to the same antenna), so I don't think we have a technical reason to choose either approach. Do you think that other applications would benefit from using to identify antennas?

**Comment by cg-laser** [comment]
I'm also surprised by your use of channel_group_id. I think it would be better if it referred to a group of antennas at the same position, i.e. dual or triple polarized antennas. 
I think it is rather a special case that all LOFAR antennas can be classified as EW and NS. To get all antennas of a specific polarization, it would be better to write a helper function of the detector class that automatically filters all antennas that have the same orientation (or rather the orientation that corresponds to a certain polarization). 

For other detectors, like the Pierre Auger Observatory, each station (=1 dual polarized antenna) has a unique index which we can model with `channel_group_id`. 

If you don't have strong feelings about the use of `channel_group_id` I suggest that you change it.

File: NuRadioReco/framework/parameters.py
Comment (MijnheerD) [reply NuRadioReco/framework/parameters.py:42]: We will change the interpretation of `channel_group_id` (as discussed), so this will be just an integer unrelated to that. I added the modules where the parameters are set/calculated.

File: NuRadioReco/modules/LOFAR/beamformingDirectionFitter_LOFAR.py
Comment (MijnheerD) [reply NuRadioReco/modules/LOFAR/beamformingDirectionFitter_LOFAR.py:17]: Could you point me to a module where this is done? I have randomly opened some modules, but none of them seem to actually implement logging as specified in documentation: https://nu-radio.github.io/NuRadioMC/NuRadioReco/pages/nur_modules.html#logging

File: NuRadioReco/modules/io/LOFAR/readLOFARData.py
Comment (MijnheerD) [reply NuRadioReco/modules/io/LOFAR/readLOFARData.py:454]: Indeed, we should add this to the Detector description (also changing the interpretation of the channel_group_id as discussed below). I'll make this a separate issue to implement in a different branch, I think it would be too messy to add this through this branch.

File: NuRadioReco/modules/io/LOFAR/rawTBBio.py
Comment (MijnheerD) [reply NuRadioReco/modules/io/LOFAR/rawTBBio.py:81]: Yeah, so this module is actually an import of a standalone library that is maintained by Brian Hare and is heavily tied to way we receive our data from ASTRON. I could change this (and many other things), but as soon as we would want to update this with a newer version of Brian's code, all of it would have to be changed again. So I prefer to keep this as is, and convert everything to the NRR unit system during read-in (where we use the functions from this module to read our data into the Event).

File: NuRadioReco/modules/LOFAR/beamformingDirectionFitter_LOFAR.py
Comment (sjoerd-bouma) [reply NuRadioReco/modules/LOFAR/beamformingDirectionFitter_LOFAR.py:143]: I don't think you've pushed your commits yet, so it's hard to tell :D 
But yes, I think my main objection was the `\` which I generally find ugly (and it seems in this case [PEP8](https://peps.python.org/pep-0008/) mostly agrees.

Comment (MijnheerD) [reply NuRadioReco/modules/LOFAR/beamformingDirectionFitter_LOFAR.py:147]: In the end I replaced this entire statement, so it's not a problem anymore ;-)

**Comment by cg-laser** [comment]
Thanks Mitja! 
I think it is good to implement the restructuring of group id's in a separate branch, but I'd rather not merge this branch into develop bevor we made the release. Otherwise, we release an unintended framework feature. 
But we could just freeze this branch, and you can create a new branch based on this one to do the group_id restructuring.

**Comment by MijnheerD** [comment]
Do you mean that you do not wish to merge the LOFAR branch into develop before the release, or the branch which implements the `group_id` in the Detector class? Or both?

**Comment by MijnheerD** [comment]
Alternatively, we have continued work on the lofar/development branch, where I already merged the Detector `channel_group_id` feature (PR #598) and parts of the logging improvements (PR #602). So if you prefer, we could also throw away this branch and simply merge the most up to date one.

**Comment by sjoerd-bouma** [comment]
@MijnheerD my preference would probably to merge #598 into this branch, so we continue to review this branch. If I understand correctly this branch contains an implementation of `channel_group_id` that we don't actually want to use and #598 rectifies this. 

I would keep the changes to the logging separate in #602. If they are already merged into #598 we should probably aim to merge #602 (separately) first.

**Comment by cg-laser** [comment]
I second Sjoerd's suggestion. We want to keep develop a fully usable version. 
#598 is difficult to review because it seems to contain all changes to this branch here as well? If this is the case, can you just make a new PR that wants to merge detector_channel_groud_id into the LOFAR branch? Then we can first review the implementation and usage of the channel_group_id. And then merge this branch here into develop. 
I think the logger changes #602 are almost done and we can merge soon. Then you can of course merge develop into this branch to keep it up to date.

**Comment by MijnheerD** [comment]
Hi @sjoerd-bouma and @cg-laser , sorry for my late reply. I created a pull request (#640) which pulls the Detector channel group ID feature (from PR #598 ) into this branch.

Concerning the logging changes, some were already pulled into this branch (through commit 8d97a0852) . This only changed one file, the base/module.py file where the handles in `setup_logger()` were changed. What do you think is the best course of action? Pulling #602 into this branch to update it with the logging changes, or first merging this branch into develop and then complete PR #602 ? (Cause I guess that doing it the other way around will result into merge conflicts)

**Comment by cg-laser** [comment]
Please merge #602 into develop, and then you can merge develop back into this branch to have it up to date.

**Comment by cg-laser** [comment]
@MijnheerD can you remind me. Does this branch define the group id such that it refers to a dual-polarized antenna? I.e., per group id, you have two antennas with orthogonal polarization?

**Comment by MijnheerD** [comment]
@cg-laser Correct, in the LOFAR modules the group ID is assumed to be the same for the two channels with have the same physical position (i.e. are part of a dual-polarised antenna). Though I think the actual implementation would also work for the case where group ID's refer to a polarisation, that depends on how one writes the JSON file.

File: NuRadioReco/framework/station.py
Comment (cg-laser) [inline NuRadioReco/framework/station.py:59]: shouldn't this function be called "remove_channel"? 

Also please add docstrings at least to the new functions. Github Copilot can often autogenerate them.

**Review by cg-laser — COMMENTED** [review]
Looks good to me except one small inline comment. Please fix this. I don't need to see the PR again. (I also can't approve it because I originally created it ;-)) So you can self-approve and merge.

File: NuRadioReco/framework/station.py
Comment (MijnheerD) [reply NuRadioReco/framework/station.py:59]: I think the reasoning here was that seeing the function accepts a channel ID rather than a Channel object, it was better named `remove_channel_id`. I have changed the function to both work with a Channel ID (ie. an integer) or a Channel object itself, and renamed to function `remove_channel`

**Review by MijnheerD — APPROVED** [review]

**Comment by fschlueter** [comment]
@MijnheerD, typically we are not approving our own PR and merge them.

**Comment by cg-laser** [comment]
I told him to do so after addressing my comment.

**Pull request merged**
